servant-foreign/src/Servant/Foreign.hs

```diff
@@ -0,0 +1,409 @@
+-- >   , requestBody = "body"
+-- >   , successCallback = "onSuccess"
+-- >   , errorCallback = "onError"
+-- >   , moduleName = ""
+-- >   , urlPrefix = ""
+-- >   }
+-- @
+defCommonGeneratorOptions :: CommonGeneratorOptions
+defCommonGeneratorOptions = CommonGeneratorOptions
+  {
+    functionNameBuilder = camelCase
+  , requestBody = "body"
+  , successCallback = "onSuccess"
+  , errorCallback = "onError"
+  , moduleName = ""
+  , urlPrefix = ""
+  }
+
+-- | Function name builder that simply concat each part together
+concatCase :: FunctionName -> String
+concatCase = concat
+
+-- | Function name builder using the snake_case convention.
+-- each part is separated by a single underscore character.
+snakeCase :: FunctionName -> String
+snakeCase = intercalate "_"
+
+-- | Function name builder using the CamelCase convention.
+-- each part begins with an upper case character.
+camelCase :: FunctionName -> String
+camelCase [] = ""
+camelCase (p:ps) = concat $ p : camelCase' ps
+   where camelCase' [] = []
+         camelCase' (r:rs) = capitalize r : camelCase' rs
+         capitalize [] = []
+         capitalize (x:xs) = toUpper x : xs
+
+type Arg = String
+
+data Segment = Segment { _segment :: SegmentType, _matrix :: [MatrixArg] }
+  deriving (Eq, Show)
+
+data SegmentType = Static String  -- ^ a static path segment. like "/foo"
+                 | Cap Arg        -- ^ a capture. like "/:userid"
+  deriving (Eq, Show)
+
+type Path = [Segment]
+
+data ArgType =
+    Normal
+  | Flag
+  | List
+  deriving (Eq, Show)
+
+data QueryArg = QueryArg
+  { _argName :: Arg
+  , _argType :: ArgType
+  } deriving (Eq, Show)
+
+data HeaderArg = HeaderArg
+    { headerArgName :: String
+    }
+  | ReplaceHeaderArg
+    { headerArgName :: String
+    , headerPattern :: String
+    } deriving (Eq)
+
+
+-- | 
+instance Show HeaderArg where
+  show (HeaderArg n)          = toValidFunctionName ("header" <> n)
+  show (ReplaceHeaderArg n p)
+      | pn `isPrefixOf` p = pv <> " + \"" <> rp <> "\""
+      | pn `isSuffixOf` p = "\"" <> rp <> "\" + " <> pv
+      | pn `isInfixOf` p  = "\"" <> (replace pn ("\" + " <> pv <> " + \"") p)
+                                 <> "\""
+      | otherwise         = p
+    where
+      pv = toValidFunctionName ("header" <> n)
+      pn = "{" <> n <> "}"
+      rp = replace pn "" p
+      -- Use replace method from Data.Text
+      replace old new = T.unpack .
+          T.replace (T.pack old) (T.pack new) .
+          T.pack
+
+-- | Attempts to reduce the function name provided to that allowed by @'Foreign'@.
+--
+-- Here we are making an assumption that js identifiers are common enough.
+-- https://mathiasbynens.be/notes/javascript-identifiers
+-- Couldn't work out how to handle zero-width characters.
+-- TODO: compare it with other generated languages(such as ruby via lackey)
+-- and generalize.
+-- 
+-- @TODO: specify better default function name, or throw error?
+toValidFunctionName :: String -> String
```

Comment:
In the end i moved the code from this instance to the `toJSHeader` function in `servant-js`. Makes more sense this way.